pkg/node/node.go

```diff
@@ -101,105 +103,261 @@ func createAddressFilter(
+// the chain. It implements retry mechanism allowing to attempt to publish again
+// in case of a failure.
+//
+// We do implement a retry in this function because the retry mechanism is much
+// more complex than in case of e.g. publishSignerPublicKey. Although all keep
+// members are supposed to try to publish the signature, only one transaction
+// succeeds. For each attempt, we need to check if the keep still awaits
+// a signature. Also, we need to implement some sane delay between attempts so
+// that we do not waste gas.
+func (n *Node) publishSignature(
+	ctx context.Context,
+	keepAddress common.Address,
+	digest [32]byte,
+	signature *ecdsa.Signature,
+) error {
+	attemptCounter := 0
+	for {
+		attemptCounter++
+
+		// Check if keep still awaits a signature for this digest.
+		// We do this check here in case the attempt was retried because of
+		// on-chain failure during submission. In this case we want to make sure
+		// no other member published the signature in the meantime so that we
+		// do not burn ether on redundant submission.
+		//
+		// If the check failed, we retry from the beginning.
 		isAwaitingSignature, err := n.ethereumChain.IsAwaitingSignature(keepAddress, digest)
 		if err != nil {
-			return fmt.Errorf("failed to verify if keep is still awaiting signature: [%v]", err)
+			logger.Errorf(
+				"failed to verify if keep [%s] is still awaiting signature: [%v]",
+				keepAddress.String(),
+				err,
+			)
+			continue
 		}
 
+		// Someone submitted the signature and it was accepted by the keep.
+		// We are fine, leaving.
 		if !isAwaitingSignature {
-			logger.Infof("signature submitted by another member: [%+x]", digest)
-
+			logger.Infof(
+				"signature for keep [%s] already submitted: [%+x]",
+				keepAddress.String(),
+				digest,
+			)
 			return nil
 		}
 
-		return fmt.Errorf("failed to submit signature: [%v]", err)
-	}
+		logger.Infof(
+			"publishing signature for keep [%s]; attempt [%v]",
+			keepAddress.String(),
+			attemptCounter,
+		)
 
-	logger.Infof("submitted signature for digest: [%+x]", digest)
+		if err := n.ethereumChain.SubmitSignature(keepAddress, signature); err != nil {
+			isAwaitingSignature, err := n.ethereumChain.IsAwaitingSignature(keepAddress, digest)
+			if err != nil {
+				logger.Errorf("failed to verify if keep is still awaiting signature: [%v]", err)
+				continue
+			}
+
+			// Check if we failed because someone else submitted in the meantime
+			// or because something wrong happened with our transaction.
+			if !isAwaitingSignature {
+				logger.Infof(
+					"signature for keep [%s] already submitted: [%+x]",
+					keepAddress.String(),
+					digest,
+				)
+				return nil
+			}
+
+			// Our public key submission transaction failed. We are going to
+			// wait for some time and then retry from the beginning.
+			logger.Errorf(
+				"failed to submit signature: [%v]; will retry after 10 minutes",
+				err,
+			)
+			time.Sleep(10 * time.Minute)
```

Comment:
Why 10 minutes? Isn’t it too much? Maybe we could monitor for an event that signature was already submitted by other member and return?

Reply:
I did not want to complicate this function even more. That's why I added this sleep and in the next iteration, we check if `IsAwaitingSignature`. 

Now when I look at it, I do agree 10 minutes is too much. On the other hand, anything below 30 seconds is probably too aggressive given the block mining times. What if I set it to 1 minute?

Reply:
> I did not want to complicate this function even more. That's why I added this sleep and in the next iteration, we check if IsAwaitingSignature.

I believe we should prefer web socket events monitoring vs contract calls to optimize infrastructure costs.

> What if we set it to 1 minute?
1 minute is a fair value.

Reply:
See https://github.com/keep-network/keep-ecdsa/pull/357/commits/c21d70d1e0b442b1eee83f02d058ca4fb3cfefae

Reply:
> I believe we should prefer web socket events monitoring vs contract calls to optimize infrastructure costs.

Agreed though I am not that concerned about this one here. This action is not executed very often and if we are looking for savings we should look them in sortition pool monitoring or minimum stake watchtower. More importantly, this one is nothing compared to the two previously mentioned and I don't have the bandwidth to refactor it right now and this is how it's in `master` so 🤷‍♂ 

Taking #364 into account, I think we should even execute `IsAwaitingSignature` another time after some delay to double-check our block was accepted 😬